src/ua_parser/re2.py

```diff
@@ -0,0 +1,80 @@
+from __future__ import annotations
+
+import io
+import os
+import re
+from typing import List, Tuple, Union
+
+import re2  # type: ignore
+
+from .core import (
+    Parser as BaseParser,
+    PartialParseResult,
+    Device,
+    Domain,
+    OS,
+    UserAgent,
+    Matchers,
+    UserAgentMatcher,
+    OSMatcher,
+    DeviceMatcher,
+)
+
+
+RE_OPTS = re2.Options()
+# as of uap-core 0.18, the devices set needs at least 28MB (up from
+# the default 8), set to 32
+RE_OPTS.max_mem = 8 << 22
+# might write directly to stdout? not great, suppress
+RE_OPTS.log_errors = False
+
+
+class Parser(BaseParser):
+    ua: re2.Set
+    user_agent_parsers: List[UserAgentMatcher]
+    os: re2.Set
+    os_parsers: List[OSMatcher]
+    devices: re2.Set
+    device_parsers: List[DeviceMatcher]
+
+    def __init__(
+        self,
+        matchers: Matchers,
+    ) -> None:
+        self.user_agent_parsers, self.os_parsers, self.device_parsers = matchers
+
+        self.ua = re2.Set.SearchSet(RE_OPTS)
+        for u in self.user_agent_parsers:
+            self.ua.Add(u.regex.pattern)
+        self.ua.Compile()
+
+        self.os = re2.Set.SearchSet(RE_OPTS)
+        for o in self.os_parsers:
+            self.os.Add(o.regex.pattern)
+        self.os.Compile()
+
+        self.devices = re2.Set.SearchSet(RE_OPTS)
+        for d in self.device_parsers:
+            # Prepend the i global flag if IGNORECASE is set. Assumes
+            # no pattern uses global flags, but since they're not
+            # supported in JS that seems safe.
+            if d.regex.flags & re.IGNORECASE:
+                self.devices.Add("(?i)" + d.regex.pattern)
+            else:
+                self.devices.Add(d.regex.pattern)
+        self.devices.Compile()
+
+    def __call__(self, ua: str, domains: Domain, /) -> PartialParseResult:
+        user_agent = os = device = None
+        if Domain.USER_AGENT in domains:
+            if matches := self.ua.Match(ua):
+                user_agent = self.user_agent_parsers[min(matches)](ua)
```

Comment:
Might be useful to leave a comment in this method: the `min` are because `re2.Set` does not return the match set in index order (might be in match position order?), so the first match is not the one we want.

Reply:
After taking a better look at the Filter exposed by Google-re2, it turns out to be trivial and not need extra dependencies but possibly less performant than hoped for as a result: the Python bindings don’t expose FilteredRE2, instead Filter implements prefiltering internally by means of a Set. 

This means it can just be swapped for Set , and will work, and might be faster than Set, but it still needs extra memory and won’t reach the heights of an aho-corasick prefilter.